orm/model_bucket.go

```diff
@@ -0,0 +1,174 @@
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is retured and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest ModelSlicePtr) error
+
+	// Put saves given model in the database. Before inserting into
+	// database, model is validated using its Validate method.
+	// If the key is nil or zero length then a sequence generator is used
+	// to create a unique key value.
+	// Using a key that alreayd exists in the database cause the value to
+	// be overwritten.
+	Put(db weave.KVStore, key []byte, m Model) ([]byte, error)
+
+	// Delete removes an entity with given primary key from the database.
+	// It returns ErrNotFound if an entity with given key does not exist.
+	Delete(db weave.KVStore, key []byte) error
+
+	// Register registers this buckets content to be accessable via query
+	// requests under the given name.
+	Register(name string, r weave.QueryRouter)
+}
+
+// NewModelBucket returns a ModelBucket instance. This implementation relies on
+// a bucket instance. Final implementation should operate directly on the
+// KVStore instead.
+func NewModelBucket(b Bucket) ModelBucket {
+	return &modelBucket{
+		b:     b,
+		idSeq: b.Sequence("id"),
+	}
+}
+
+type modelBucket struct {
+	b     Bucket
+	idSeq Sequence
+}
+
+func (mb *modelBucket) Register(name string, r weave.QueryRouter) {
+	mb.b.Register(name, r)
+}
+
+func (mb *modelBucket) One(db weave.ReadOnlyKVStore, key []byte, dest Model) error {
+	obj, err := mb.b.Get(db, key)
+	if err != nil {
+		return err
+	}
+	if obj == nil || obj.Value() == nil {
+		return errors.Wrapf(errors.ErrNotFound, "%T not in the store", dest)
+	}
+	res := obj.Value()
+
+	if !reflect.TypeOf(res).AssignableTo(reflect.TypeOf(dest)) {
+		return errors.Wrapf(errors.ErrType, "%T cannot be represented as %T", res, dest)
+	}
+
+	reflect.ValueOf(dest).Elem().Set(reflect.ValueOf(res).Elem())
+	return nil
+}
+
+func (mb *modelBucket) ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, destination ModelSlicePtr) error {
+	objs, err := mb.b.GetIndexed(db, indexName, key)
+	if err != nil {
+		return err
+	}
+	if len(objs) == 0 {
+		return nil
+	}
+
+	dest := reflect.ValueOf(destination)
+	if dest.Kind() != reflect.Ptr {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of models")
+	}
+	if dest.IsNil() {
+		return errors.Wrap(errors.ErrImmutable, "got nil pointer")
+	}
+	dest = dest.Elem()
+	if dest.Kind() != reflect.Slice {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of models")
+	}
+
+	// It is allowed to pass destination as both []MyModel and []*MyModel
+	sliceOfPointers := dest.Type().Elem().Kind() == reflect.Ptr
+
+	for _, obj := range objs {
+		if obj == nil || obj.Value() == nil {
+			continue
+		}
+		val := reflect.ValueOf(obj.Value())
+		if !sliceOfPointers {
+			val = val.Elem()
+		}
+		dest.Set(reflect.Append(dest, val))
```

Comment:
We loose access to `obj.Key()` here, right?
This is actual essential info in many places, eg. if the ID is username, and not stored in the Value.

Either we need to rethink that and include the Key inside the serialized Value everywhere (space overhead), or come up with a more clever form to cast them.

eg. dst could expect `[]struct{Key []byte, Value <value type>}` instead of `[]<value type>`.
Other ideas?

Reply:
There is a better option for this, we could return idSlice() as a result of this method too, with matching indices and then use helper to match model -> key

Reply:
so it would look like:

```
ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest orm.ModelSlicePtr) (ids [][]byte, error)
```

Reply:
It's much better than passing in a struct, because it removes the boilerplate.
Since the basic idea of the package is to remove as much boilerplate from working with a Bucket as possible - I suppose it's crucial to avoid introducing it.

Reply:
I like @ruseinov proposal of adding another return parameter.

I have greped through the code and there are only two places that use secondary index:
1. `./x/namecoin/wallet.go` use it to provide an API that no one is using
2. `./x/gov/bucket.go` use it instead of the main index (why?)

In the first case, this is a dead API that no one is using.
In the second case, I don't understand why secondary index is used in the first place. There are two extra indexes declared (3 in total) and one is ignored.

I could not find any existing code where the ID of a group of items is needed. If we really do need it, than I think what Roman proposed is the best.

Reply:
@ethanfrey ^^

Reply:
I have extended the `ByIndex` method to return keys as well in https://github.com/iov-one/weave/pull/615/commits/0181a82d725603722238eba340295e035fabc4ee :man_shrugging:

Reply:
Very nice API update here, and great suggestion @ruseinov